server/schedulers/grant_leader.go

```diff
@@ -102,26 +170,144 @@ func (s *grantLeaderScheduler) EncodeConfig() ([]byte, error) {
+	// return []*operator.Operator{op}
+
 	schedulerCounter.WithLabelValues(s.GetName(), "schedule").Inc()
-	region := cluster.RandFollowerRegion(s.conf.StoreID, s.conf.Ranges, opt.HealthRegion(cluster))
-	if region == nil {
-		schedulerCounter.WithLabelValues(s.GetName(), "no-follower").Inc()
-		return nil
-	}
-	schedulerCounter.WithLabelValues(s.GetName(), "new-operator").Inc()
-	op := operator.CreateTransferLeaderOperator(GrantLeaderType, region, region.GetLeader().GetStoreId(), s.conf.StoreID, operator.OpLeader)
-	op.SetPriorityLevel(core.HighPriority)
-	return []*operator.Operator{op}
+	var ops []*operator.Operator
+	s.conf.mu.RLock()
+	defer s.conf.mu.RUnlock()
+	for id, ranges := range s.conf.StoreIDWitRanges {
+		region := cluster.RandFollowerRegion(id, ranges, opt.HealthRegion(cluster))
+		if region == nil {
+			schedulerCounter.WithLabelValues(s.GetName(), "no-follower").Inc()
+			continue
+		}
+
+		schedulerCounter.WithLabelValues(s.GetName(), "new-operator").Inc()
+		op := operator.CreateTransferLeaderOperator(GrantLeaderType, region, region.GetLeader().GetStoreId(), id, operator.OpLeader)
+		op.SetPriorityLevel(core.HighPriority)
+		ops = append(ops, op)
+	}
+
+	return ops
+}
+
+type grantLeaderHandler struct {
+	rd     *render.Render
+	config *grantLeaderSchedulerConfig
+}
+
+func (handler *grantLeaderHandler) UpdateConfig(w http.ResponseWriter, r *http.Request) {
+	var input map[string]interface{}
+	if err := apiutil.ReadJSONRespondError(handler.rd, w, r.Body, &input); err != nil {
+		return
+	}
+	var args []string
+	var exists bool
+	var id uint64
+	idFloat, ok := input["store_id"].(float64)
+	if ok {
+		id = (uint64)(idFloat)
+		if _, exists = handler.config.StoreIDWitRanges[id]; !exists {
+			if err := handler.config.cluster.BlockStore(id); err != nil {
+				handler.rd.JSON(w, http.StatusInternalServerError, err)
+				return
+			}
+		}
+		args = append(args, strconv.FormatUint(id, 10))
+	}
+
+	ranges, ok := (input["ranges"]).([]string)
+	if ok {
+		args = append(args, ranges...)
+	} else if exists {
+		args = append(args, handler.config.getRanges(id)...)
+	}
+
+	handler.config.BuildWithArgs(args)
+	err := handler.config.Persist()
+	if err != nil {
+		handler.rd.JSON(w, http.StatusInternalServerError, err)
+	}
+	handler.rd.JSON(w, http.StatusOK, nil)
+}
+
+func (handler *grantLeaderHandler) ListConfig(w http.ResponseWriter, r *http.Request) {
+	conf := handler.config.Clone()
+	handler.rd.JSON(w, http.StatusOK, conf)
+}
+
+func (handler *grantLeaderHandler) DeleteConfig(w http.ResponseWriter, r *http.Request) {
+	idStr := mux.Vars(r)["store_id"]
+	id, err := strconv.ParseUint(idStr, 10, 64)
+	if err != nil {
+		handler.rd.JSON(w, http.StatusBadRequest, err.Error())
+		return
+	}
+
+	handler.config.mu.Lock()
+	defer handler.config.mu.Unlock()
+	_, exists := handler.config.StoreIDWitRanges[id]
+	if exists {
+		delete(handler.config.StoreIDWitRanges, id)
+		handler.config.cluster.UnblockStore(id)
+
+		handler.config.mu.Unlock()
```

Comment:
can add a common function that not use `mu` to `persist`,  then can remove this strange code.